kratos/includes/global_pointer.h

```diff
@@ -22,272 +22,409 @@
-   * @param rOther The GlobalPointer object to be compared.
-   * @return true if the underlying pointers are equal, false otherwise.
-   */
-  bool operator==(const GlobalPointer& rOther)
-  {
+    /**
+     * @brief Overloads the '==' operator to compare two GlobalPointer objects of the same template type. Returns true if the underlying pointers are equal.
+     * @param rOther The GlobalPointer object to be compared.
+     * @return true if the underlying pointers are equal, false otherwise.
+     */
+    bool operator==(const GlobalPointer& rOther)
+    {
 #ifdef KRATOS_USING_MPI
-    return this->get() == rOther.get() && this->GetRank() == rOther.GetRank();
-#else 
-    return this->get() == rOther.get();
+      return this->get() == rOther.get() && this->GetRank() == rOther.GetRank();
+#else
+      return this->get() == rOther.get();
 #endif
-  }
-
-  /** Fills buffer with the GlobalPoiter data
-   * Fills buffer with the GlobalPoiter data
-   * @param buffer Object data buffer
-   */
-  void save(char * buffer) const {
-    memcpy(buffer, this, sizeof(GlobalPointer));
-  }
-
-  /** Restores the GlobalPoiter with the data from the buffer
-   * Restores the GlobalPoiter with the data from the buffer
-   * @param buffer Object data buffer
-   */
-  void load(char * buffer) {
-    memcpy(this, buffer, sizeof(GlobalPointer));
-  }
-
-  /** Returns the rank of the global pointer
-   * Returns the rank of the global pointer data is located or 0 if no mpi
-   * @return rank of the global pointer data or 0
-   */
-  int GetRank() const {
+    }
+
+    ///@}
+    ///@name Access
+    ///@{
+
+    /**
+    * @brief Get a non-const pointer to the data.
+    * @return A pointer to the data.
+    */
+    TDataType* get() {
+        return mDataPointer;
+    }
+
+    /**
+    * @brief Get a const pointer to the data.
+    * @return A const pointer to the data.
+    */
+    TDataType const* get() const {
+        return mDataPointer;
+    }
+
+    /**
+     * @brief Returns the rank of the global pointer
+     * @details Returns the rank of the global pointer data is located or 0 if no mpi
+     * @return rank of the global pointer data or 0
+     */
+    int GetRank() const
+    {
 #ifdef KRATOS_USING_MPI
-    return this->mRank;
+      return this->mRank;
 #else
-    return 0;
+      return 0;
 #endif
-  }
-
-  private:
-
-  friend class Serializer;
-
-  void save(Serializer& rSerializer) const
-  {
-      if(rSerializer.Is(Serializer::SHALLOW_GLOBAL_POINTERS_SERIALIZATION))
-      {
-        rSerializer.save("D", reinterpret_cast<std::size_t>(mDataPointer));
-      }
-      else
-      {
-        rSerializer.save("D", mDataPointer);
-      }
+    }
+
+    /**
+     * @brief Sets the rank of the global pointer
+     * @details Sets the rank of the global pointer where data is located
+     * @param Rank The rank to set
+     */
+    void SetRank(const int Rank)
+    {
 #ifdef KRATOS_USING_MPI
```

Comment:
well, i guess we should have a debug error here in case we are not in mpi mode

Reply:
And rank is different than 0, because if you implement a generic code in serial it will set always to 0, otherwise you will create many false errors

Reply:
well, imagine you try to set something to rank 5 in serial... that would be silently ignored